src/Raven.Server/Smuggler/Migration/Migrator.cs

```diff
@@ -0,0 +1,353 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Session;
+using Raven.Client.Documents.Smuggler;
+using Raven.Client.ServerWide;
+using Raven.Server.Documents;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.Operations;
+using Raven.Server.Json;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Smuggler.Documents;
+using Raven.Server.Web.System;
+using Sparrow.Json;
+using DatabaseSmuggler = Raven.Server.Smuggler.Documents.DatabaseSmuggler;
+
+namespace Raven.Server.Smuggler.Migration
+{
+    public class Migrator
+    {
+        private const string MigrationStateKeyBase = "Raven/Migration/Status";
+
+        private readonly HttpClient _client;
+        private readonly string _serverUrl;
+        private readonly string _migrationStateKey;
+        private readonly ServerStore _serverStore;
+        private readonly Operations _operations;
+        private readonly CancellationToken _cancellationToken;
+
+        public Migrator(
+            MigrationConfigurationBase configuration, 
+            ServerStore serverStore,
+            Operations operations,
+            CancellationToken cancellationToken)
+        {
+            var hasCredentials =
+                string.IsNullOrWhiteSpace(configuration.UserName) == false &&
+                string.IsNullOrWhiteSpace(configuration.Password) == false;
+
+            var httpClientHandler = new HttpClientHandler
+            {
+                UseDefaultCredentials = hasCredentials == false,
+            };
+            if (hasCredentials)
+                httpClientHandler.Credentials = new NetworkCredential(
+                    configuration.UserName,
+                    configuration.Password,
+                    configuration.Domain);
+
+            _client = new HttpClient(httpClientHandler);
+            while (configuration.ServerUrl.EndsWith("/"))
+                configuration.ServerUrl = configuration.ServerUrl.Remove(configuration.ServerUrl.Length - 1);
+
+            _serverUrl = configuration.ServerUrl.ToLower();
```

Comment:
Why `ToLower` ?

Reply:
I'm saving and loading a document (save the last etag from the destination) using the server url.
The document ids are non case sensitive so this can be removed.